crates/ruff_python_parser/src/lexer.rs

```diff
@@ -847,39 +999,66 @@ impl<'source> Lexer<'source> {
                 if self.cursor.eat_char('=') {
                     Tok::NotEqual
                 } else {
-                    return Err(LexicalError {
-                        error: LexicalErrorType::UnrecognizedToken { tok: '!' },
-                        location: self.token_start(),
-                    });
+                    Tok::Exclamation
                 }
             }
             '~' => Tok::Tilde,
             '(' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    fstring_context.open_parentheses();
+                }
                 self.nesting += 1;
                 Tok::Lpar
             }
             ')' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    fstring_context.close_parentheses();
+                }
                 self.nesting = self.nesting.saturating_sub(1);
                 Tok::Rpar
             }
             '[' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    fstring_context.open_parentheses();
+                }
                 self.nesting += 1;
                 Tok::Lsqb
             }
             ']' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    fstring_context.close_parentheses();
+                }
                 self.nesting = self.nesting.saturating_sub(1);
                 Tok::Rsqb
             }
             '{' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    fstring_context.open_parentheses();
+                }
                 self.nesting += 1;
                 Tok::Lbrace
             }
             '}' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    if !fstring_context.has_open_parentheses() {
+                        return Err(LexicalError {
+                            error: LexicalErrorType::FStringError(FStringErrorType::SingleRbrace),
+                            location: self.token_start(),
+                        });
+                    }
+                    fstring_context.close_parentheses();
+                }
                 self.nesting = self.nesting.saturating_sub(1);
                 Tok::Rbrace
             }
             ':' => {
-                if self.cursor.eat_char('=') {
+                if self
+                    .fstring_stack
+                    .last_mut()
+                    .is_some_and(FStringContext::is_valid_format_spec_position)
+                {
+                    Tok::Colon
+                } else if self.cursor.eat_char('=') {
```

Comment:
Is this because `:=` has a different meaning inside of a format spec?

Reply:
Does https://github.com/astral-sh/ruff/pull/6659#discussion_r1303742193 answer your question here?